### PR TITLE
Pyright fixes for image drivers and image model drivers

### PR DESCRIPTION
**Pull request by andrewfrench**
Fixes a chunk of #519

### CODE REVIEW

File: griptape/drivers/image_generation/amazon_bedrock_image_generation_driver.py [hunk 1]
```diff
@@ -1,16 +1,15 @@
...
 from attr import define, field, Factory
-
 from griptape.artifacts import ImageArtifact
 from griptape.drivers import BaseMultiModelImageGenerationDriver
+
 from griptape.utils import import_optional_dependency
 
 if TYPE_CHECKING:
-    import boto3
+    import boto3  # type: ignore
```
Comment (andrewfrench) [inline griptape/drivers/image_generation/amazon_bedrock_image_generation_driver.py:12]: `import boto3` leaves us with this error:


> Stub file not found for "boto3"Pylance[reportMissingTypeStubs](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportMissingTypeStubs)

Very annoying, but it looks like it can be resolved by [boto3-stubs](https://pypi.org/project/boto3-stubs). Ignoring for now until we can play around with it.

Comment (collindutter) [reply griptape/drivers/image_generation/amazon_bedrock_image_generation_driver.py:12]: Looks like that package is just an install and we're good? Should we just knock that out now?

File: griptape/drivers/image_generation/openai_dalle_image_generation_driver.py [hunk 2]
```diff
@@ -48,13 +50,12 @@ class OpenAiDalleImageGenerationDriver(BaseImageGenerationDriver):
     def try_generate_image(self, prompts: list[str], negative_prompts: list[str] | None = None) -> ImageArtifact:
         prompt = ", ".join(prompts)
 
-        additional_params = {}
+        additional_params = dict[str, Any]()
```
Comment (collindutter) [inline griptape/drivers/image_generation/openai_dalle_image_generation_driver.py:53]: I think we use the alternate syntax in other places. I think it's a bit easier to read.
```python
additional_params: dict[str, Any] = {}
```

File: griptape/drivers/image_generation_model/amazon_bedrock_stable_diffusion_image_generation_model_driver.py [hunk 3]
```diff
@@ -21,7 +21,7 @@ def text_to_image_request_parameters(
         image_height: int,
         negative_prompts: Optional[list[str]] = None,
         seed: Optional[int] = None,
-    ) -> dict:
+    ) -> Dict[str, Any]:
```
Comment (collindutter) [inline griptape/drivers/image_generation_model/amazon_bedrock_stable_diffusion_image_generation_model_driver.py:24]: Don't need to use `Typing.Dict` beyond python 3.7 (might be off there). Can replace with `dict[str, Any]`.

File: griptape/drivers/image_generation_model/amazon_bedrock_stable_diffusion_image_generation_model_driver.py [hunk 4]
```diff
@@ -46,7 +46,7 @@ def text_to_image_request_parameters(
 
         return request
 
-    def get_generated_image(self, response: dict) -> bytes:
+    def get_generated_image(self, response: Dict[str, Any]) -> bytes:
```
Comment (collindutter) [inline griptape/drivers/image_generation_model/amazon_bedrock_stable_diffusion_image_generation_model_driver.py:49]: Same

File: griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py [hunk 5]
```diff
@@ -19,30 +19,32 @@ def text_to_image_request_parameters(
         image_height: int,
         negative_prompts: Optional[list[str]] = None,
         seed: Optional[int] = None,
-    ) -> dict:
+    ) -> Dict[str, Any]:
```
Comment (collindutter) [inline griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py:22]: Replace with `dict[str, Any]`

File: griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py [hunk 6]
```diff
@@ -19,30 +19,32 @@ def text_to_image_request_parameters(
...
-            "textToImageParams": {"text": prompt},
-            "imageGenerationConfig": {
-                "numberOfImages": 1,
-                "quality": self.quality,
-                "width": image_width,
-                "height": image_height,
-                "cfgScale": self.cfg_scale,
-            },
-        }
+        request = dict[str, Any]()
```
Comment (collindutter) [inline griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py:25]: Use other syntax

File: griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py [hunk 7]
```diff
@@ -19,30 +19,32 @@ def text_to_image_request_parameters(
...
-            "taskType": self.task_type,
-            "textToImageParams": {"text": prompt},
-            "imageGenerationConfig": {
-                "numberOfImages": 1,
-                "quality": self.quality,
-                "width": image_width,
-                "height": image_height,
-                "cfgScale": self.cfg_scale,
-            },
-        }
```
Comment (collindutter) [inline griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py:35]: Building the dict with all its keys at once feels cleaner than building it incrementally. Why the change?

File: griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py [hunk 8]
```diff
@@ -19,30 +19,32 @@ def text_to_image_request_parameters(
...
+        request = dict[str, Any]()
+        request["taskType"] = self.task_type
 
+        text_to_image_params = dict[str, Any]()
```
Comment (collindutter) [inline griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py:28]: Other syntax

File: griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py [hunk 9]
```diff
@@ -19,30 +19,32 @@ def text_to_image_request_parameters(
...
+        text_to_image_params["text"] = prompt
         if negative_prompts:
-            request["textToImageParams"]["negativeText"] = ", ".join(negative_prompts)
-
+            text_to_image_params["negativeText"] = ", ".join(negative_prompts)
+
+        image_generation_config = dict[str, Any]()
```
Comment (collindutter) [inline griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py:33]: Other syntax

File: griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py [hunk 10]
```diff
@@ -19,30 +19,32 @@ def text_to_image_request_parameters(
...
-            request["imageGenerationConfig"]["seed"] = seed
+            image_generation_config["seed"] = seed
+
+        request["textToImageParams"] = text_to_image_params
+        request["imageGenerationConfig"] = image_generation_config
 
         return request
 
-    def get_generated_image(self, response: dict) -> bytes:
+    def get_generated_image(self, response: Dict[str, Any]) -> bytes:
```
Comment (collindutter) [inline griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py:47]: `dict`

File: griptape/drivers/image_generation_model/base_image_generation_model_driver.py [hunk 11]
```diff
@@ -14,9 +14,9 @@ def text_to_image_request_parameters(
         image_height: int,
         negative_prompts: Optional[list[str]] = None,
         seed: Optional[int] = None,
-    ) -> dict:
+    ) -> Dict[str, Any]:
         ...
 
     @abstractmethod
-    def get_generated_image(self, response: dict) -> bytes:
+    def get_generated_image(self, response: Dict[str, Any]) -> bytes:
```
Comment (collindutter) [inline griptape/drivers/image_generation_model/base_image_generation_model_driver.py:21]: `dict`

File: tests/unit/drivers/image_generation/test_azure_openai_dalle_image_generation_driver.py [hunk 12]
```diff
@@ -16,23 +14,23 @@ def driver(self):
...
-    def test_init(self, driver):
+    def test_init(self, driver: AzureOpenAiDalleImageGenerationDriver):
         assert driver
 
     def test_init_requires_endpoint(self):
         with pytest.raises(TypeError):
             AzureOpenAiDalleImageGenerationDriver(
                 model="dall-e-3", client=Mock(), azure_deployment="dalle-deployment", image_size="512x512"
-            )
+            )  # type: ignore
```
Comment (collindutter) [inline tests/unit/drivers/image_generation/test_azure_openai_dalle_image_generation_driver.py:24]: Why the ignore?

File: tests/unit/drivers/image_generation/test_azure_openai_dalle_image_generation_driver.py [hunk 13]
```diff
@@ -16,23 +14,23 @@ def driver(self):
...
 
     def test_init_requires_deployment(self):
         with pytest.raises(TypeError):
             AzureOpenAiDalleImageGenerationDriver(
                 model="dall-e-3", client=Mock(), azure_endpoint="https://dalle.example.com", image_size="512x512"
-            )
+            )  # type: ignore
```
Comment (collindutter) [inline tests/unit/drivers/image_generation/test_azure_openai_dalle_image_generation_driver.py:30]: Why the ignore?

**Review by collindutter — CHANGES_REQUESTED** [review]

File: tests/unit/drivers/image_generation/test_azure_openai_dalle_image_generation_driver.py (hunk 12 above)
Comment (andrewfrench) [reply tests/unit/drivers/image_generation/test_azure_openai_dalle_image_generation_driver.py:24]: We're intentionally failing to provide some fields to test validation, but pyright doesn't know that.

File: griptape/drivers/image_generation/amazon_bedrock_image_generation_driver.py (hunk 1 above)
Comment (andrewfrench) [reply griptape/drivers/image_generation/amazon_bedrock_image_generation_driver.py]: Done

File: tests/unit/drivers/image_generation/test_leonardo_image_generation_driver.py [hunk 14]
```diff
@@ -6,58 +10,58 @@
...
-    def driver(self):
+    def driver(self) -> LeonardoImageGenerationDriver:
         requests_session = Mock()
 
         return LeonardoImageGenerationDriver(
             api_key="test_api_key", requests_session=requests_session, model="test_model_id"
         )
 
-    def test_init(self, driver):
+    def test_init(self, driver: LeonardoImageGenerationDriver) -> None:
```
[the remark this answers is not in the document]
Comment (andrewfrench) [reply tests/unit/drivers/image_generation/test_leonardo_image_generation_driver.py:20]: Yes, changes here are only addressing Pyright complaints. The input argument needs a type annotation, but then the return value mock doesn't align with the annotated type. This is why the mock setup on the next line in these tests is ignored. (re: [this comment](https://github.com/griptape-ai/griptape/pull/521#discussion_r1434276748))

File: griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py (hunk 7 above)
Comment (andrewfrench) [reply griptape/drivers/image_generation_model/amazon_bedrock_titan_image_generation_model_driver.py:35]: Each nested dict needs to be typed and I hadn't used the `request: dict[str, Any] = {...}` syntax. Updated to use the literal syntax but nested dicts still must be typed.

File: pyproject.toml [hunk 15]
```diff
@@ -53,6 +53,7 @@ opensearch-py = { version = "^2.3.1", optional = true }
 requests-aws4auth = { version = "1.2.3", optional = true }
 pgvector = { version = "^0.2.3", optional = true }
 psycopg2-binary = { version = "^2.9.9", optional = true }
+boto3-stubs = { version = "^1.34.5", optional = true }
```
Comment (collindutter) [inline pyproject.toml:56]: This should be added to the `dev` group using the `--with dev` flag.
